src/storage/src/sink/iceberg.rs

```diff
@@ -0,0 +1,1487 @@
+                return Err(anyhow::anyhow!(
+                    "Iceberg table '{}' partition spec changed during operation but partition spec evolution isn't supported, expected partition spec ID {}, got {}",
+                    table_name,
+                    current_partition_spec,
+                    reloaded_partition_spec
+                ));
+            }
+
+            Ok(table)
+        }
+        Err(err) => Err(err).context("Failed to reload Iceberg table"),
+    }
+}
+
+/// Load an existing Iceberg table or create it if it doesn't exist.
+async fn load_or_create_table(
+    catalog: &dyn Catalog,
+    namespace: String,
+    table_name: String,
+    schema: &Schema,
+) -> anyhow::Result<iceberg::table::Table> {
+    let namespace_ident = NamespaceIdent::new(namespace.clone());
+    let table_ident = TableIdent::new(namespace_ident.clone(), table_name.clone());
+
+    // Try to load the table first
+    match catalog.load_table(&table_ident).await {
+        Ok(table) => {
+            // Table exists, return it
+            // TODO: Add proper schema evolution/validation to ensure compatibility
+            Ok(table)
+        }
+        Err(err) => {
+            if matches!(err.kind(), ErrorKind::TableNotFound { .. }) {
+                // Table doesn't exist, create it
+                // Note: location is not specified, letting the catalog determine the default location
+                // based on its warehouse configuration
+                let table_creation = TableCreation::builder()
+                    .name(table_name.clone())
+                    .schema(schema.clone())
+                    // Use unpartitioned spec by default
+                    // TODO: Consider making partition spec configurable
+                    // .partition_spec(UnboundPartitionSpec::builder().build())
+                    .build();
+
+                catalog
+                    .create_table(&namespace_ident, table_creation)
+                    .await
+                    .with_context(|| {
+                        format!(
+                            "Failed to create Iceberg table '{}' in namespace '{}'",
+                            table_name, namespace
+                        )
+                    })
+            } else {
+                // Some other error occurred
+                Err(err).context("Failed to load Iceberg table")
+            }
+        }
+    }
+}
+
+/// Find the most recent Materialize frontier from Iceberg snapshots.
+/// We store the frontier in snapshot metadata to track where we left off after restarts.
+/// Snapshots with operation="replace" (compactions) don't have our metadata and are skipped.
+fn retrieve_upper_from_snapshots(
+    snapshots: impl IntoIterator<Item = Arc<Snapshot>>,
+) -> anyhow::Result<Option<(Antichain<Timestamp>, Option<u64>)>> {
+    let snapshots = snapshots
+        .into_iter()
+        .sorted_by(|a, b| Ord::cmp(&b.sequence_number(), &a.sequence_number()));
```

Comment:
All callers of this function seem to be doing this sort outside, and the `sorted_by` iterator combinator needs to collect into a temp vector internally. It's probably worth switching this function to one that accepts a `&[Arc<Snapshot>]` or a `&mut [Arc<Snapshot>]` if you want to sort defensively inside